tests/0101-fetch-from-follower.cpp

```diff
@@ -0,0 +1,305 @@
+  std::string errstr;
+  RdKafka::ErrorCode err;
+  class RdKafka::Metadata *metadata;
+
+  /* Determine the ids of the brokers that the partition has replicas
+   * on and which one of those is the leader.
+   */
+  RdKafka::Conf *pConf;
+  Test::conf_init(&pConf, NULL, 10);
+  RdKafka::Producer *p = RdKafka::Producer::create(pConf, errstr);
+  delete pConf;
+  test_assert(p, tostr() << "Failed to create producer: " << errstr);
+
+  RdKafka::Conf *tconf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);
+  RdKafka::Topic *topic = RdKafka::Topic::create(p, topic_str, tconf, errstr);
+  delete tconf;
+  test_assert(topic, tostr() << "Failed to create topic: " << errstr);
+
+  err = p->metadata(0, topic, &metadata, 5000);
+  test_assert(err == RdKafka::ERR_NO_ERROR,
+    tostr() <<  "%% Failed to acquire metadata: "
+            << RdKafka::err2str(err));
+
+  test_assert(metadata->topics()->size() == 1,
+    tostr() << "expecting metadata for exactly one topic. "
+            << "have metadata for " << metadata->topics()->size()
+            << "topics");
+
+  RdKafka::Metadata::TopicMetadataIterator topicMetadata = metadata->topics()->begin();
+  RdKafka::TopicMetadata::PartitionMetadataIterator partitionMetadata = (*topicMetadata)->partitions()->begin();
+
+  *leader = (*partitionMetadata)->leader();
+
+  size_t idx = 0;
+  RdKafka::PartitionMetadata::ReplicasIterator replicasIterator;
+  for (replicasIterator = (*partitionMetadata)->replicas()->begin();
+           replicasIterator != (*partitionMetadata)->replicas()->end();
+           ++replicasIterator) {
+    brokers.push_back(*replicasIterator);
+    idx++;
+  }
+
+  delete metadata;
+  delete topic;
+  delete p;
+}
+
+
+
+static void do_fff_test (void) {
+
+  /* Produce some messages to a single partition topic
+   * with 3 replicas.
+   */
+  int msgcnt = 1000;
+  const int msgsize = 100;
+  std::string topic_str = Test::mk_topic_name("0101-fetch-from-follower", 1);
+  test_create_topic(NULL, topic_str.c_str(), 1, 3);
+  test_produce_msgs_easy_size(topic_str.c_str(), 0, 0, msgcnt, msgsize);
+
+  int leader_id;
+  std::vector<int> replica_ids;
+  get_brokers_info(topic_str, &leader_id, replica_ids);
+  test_assert(replica_ids.size() == 3, tostr() << "expecting three replicas, but " << replica_ids.size() << " were reported.");
+  Test::Say(tostr() << topic_str << " leader id: " << leader_id << ", all replica ids: [" << replica_ids[0] << ", " << replica_ids[1] << ", " << replica_ids[2] << "]\n");
+
+  /* arrange for the consumer's client.rack to align with a broker that is not the leader. */
+  int client_rack_id = -1;
+  size_t i;
+  for (i=0; i<replica_ids.size(); ++i) {
+    if (replica_ids[i] != leader_id) {
+      client_rack_id = replica_ids[i];
+      break;
+    }
+  }
+  std::string client_rack = tostr() << "RACK" << client_rack_id;
+  Test::Say(tostr() << "client.rack: " << client_rack << "\n");
```

Comment:
nit: dont need tostr() stream when it is all strings, can use + string concat instead.

Reply:
yeah, i know. i find it interesting how many elementary mistakes i'm making. i think because little bits of mental effort add up to a lot over many places.